Custom ifp animations

saml1er opened the pull request:
EDIT: Github automatically closed my PR while I was trying to fix it, so I'm creating a new one.

I've implemented custom IFP animations into MTA while touching the current animation system as little as possible. I've written my own IFP loader that will allow scripters to load their own IFP file using engineLoadIFP and play animations using setPedAnimation, or replace and restore them using engineReplaceAnimation and engineRestoreAnimation respectively. The code is client-side only. 

There are three new Lua functions added:
```
// loads IFP with a custom block name 
ifp engineLoadIFP ( string ifp_file_path, string custom_block_name ) 

// this will replace an internal GTA animation with custom one, it's a light-weight operation,
// don't worry too much about performance. Different peds can have different running, walking,
// crouching, shooting, etc. animations all running simultaneously because we are not actually
// replacing animation hierarchies, we are merely storing everything in std::map which is in
// CClientPed. When an animation triggers, we get the ped by clump, and play the animation
// we wish to play. setPedAnimation works a little different than this, but the idea is same.
bool engineReplaceAnimation ( ped thePed, string block_name, string anim_name, string custom_block_name, string custom_anim_name )

// This will restore animation replaced using engineReplaceAnimation, if only 1 parameter 
// is provided which is ped, then it will restore all animations, if  block name is also provided,
// then it will restore animations of that block only, if animation name is provided as well,
// it will restore that specific animation only.
bool engineRestoreAnimation ( ped thePed [, string block_name, string anim_name ] )
```
To unload IFP, use destroyElement, and to play the custom animation simply use setPedAnimation. Also, I've tested everything countless times, but if you find something that is not the way it should be then please let me know. 

This is stable because animation loading, playing, and unloading is handled by my custom IFP loader, so we know exactly how it works. I did not use any hackish techniques in my code. I've tried my best sticking to KISS principle, so anyone can maintain it.

codenulls committed:
Displays Console. I'm only using it for testing

codenulls committed:
added a new function "CopyAnimation."

codenulls committed:
added CopyAnimation function

codenulls committed:
changed return type of AddAnimationHandler to CAnimBlendAssociation *

codenulls committed:
Animation is copied using CopyAnimation function and the result is returned by AddAnimationHandler

codenulls committed:
added CopyAnimation function

codenulls committed:
added typedef of CAnimBlendAssocGroup::CopyAnimation and changed return type of StaticAddAnimationHandler and AddAnimationHandler to CAnimBlendAssociation *

codenulls committed:
changed CAnimBlendAssociation * to CAnimBlendAssociationSAInterface *

codenulls committed:
Aligned all function prototypes properly to match StaticAddAnimationHandler prototype

codenulls committed:
Modified AddAnimation hook to skip CopyAnimation function call, if m_pAddanimationHandler is nullptr/NULL then normal function flow executes to avoid crash

codenulls committed:
removed 2 instructions in HOOK_CAnimManager_AddAnimation, because they were not needed

codenulls committed:
removed a space, nothing else

codenulls committed:
added CreateAnimBlendAssocGroup and getAnimAssocGroupInterface functions

codenulls committed:
added CreateAnimBlendAssocGroup function, also fixed tiny bits of formatting

codenulls committed:
Using factory method for access association group, instead of ugly direct memory access

codenulls committed:
Removed hCAnimBlendAssocGroup_CopyAnimation typedef

codenulls committed:
added 3 new functions for finding the correct CClientPed * in a map of  Ped clumps.

codenulls committed:
In ModelRequestCallback, old clump is removed from CAnimManagerSA::m_mapOfPedClumps and new clump  is inserted into the map

codenulls committed:
correct CClientPed * is returned from GetClientPedFromClumpMap to check to whom the clump belongs and what animation to play, it can be a custom animation as well.

codenulls committed:
added IFP entity

codenulls committed:
added scripting function engineLoadIFP for loading IFP on client.
CClientIFP::LoadIFP is empty for now.

codenulls committed:
added additional parameter to engineLoadIFP for custom block name

codenulls committed:
added CClientIFP class, but the functions in there are pretty much empty

codenulls committed:
changed iHashKey type from int to unsigned int since it's generated using CKeyGen::GetUpperCaseKey

codenulls committed:
completed CAnimBlendStaticAssociationSAInterface class members by adding missing vTable, nNumBlendNodes, and pAnimBlendNodesSequeceArray

codenulls committed:
added isGateWayAnimationHierarchy prototype to class

codenulls committed:
fixed GetAnimAssociation function and renamed it to GetAnimstaticAssociation, also implemented isGateWayAnimationHierarchy

codenulls committed:
added GetAnimStaticAssociation prototype to CAnimManagerSA class and added gateway animation constant, value is set to "run_wuzi" for gateway animation

codenulls committed:
[NOT TESTED] isGateWayAnimationHierarchy is called in AddAnimationHandler for checking and playing custom animation, but it is not tested because IFP reader needs to be added asap.

codenulls committed:
moved GATEWAY_ANIMATION_NAME from global space into CAnimManagerSA class. Also, renamed it to m_kGateWayAnimationName.

codenulls committed:
commented printf in CClient::LoadIFP and CClient::UnloadIFP

codenulls committed:
[BAD DESIGN] added IFP loader. It works properly, but the design is horrible. We'll need to properly rewrite it according to MTA's coding guidelines. Also, there is no function for unloading IFP, so yeah there are memory leaks. Dynamic memory allocation is used for sequeneces and key frames only, so freeing this memory is important.

codenulls committed:
fixed private variables

codenulls committed:
[IMPORTANT] Fixed heap corruption by adding int32_t to IFP_ANIM struct and fixed IFP_Animation struct.

codenulls committed:
[BAD DESIGN] Implemented temporary UnloadIFP function for unloading IFPs while testing. Also, using malloc instead of OLD_CMemoryMgr_Malloc.

codenulls committed:
[FIX] Memory was not being freed properly for IFP Version 2, fixed it by adding a check. Also the loop for sequences in unloadIFP was incorrect, fixed that as well. Tested by checking the pointers and frees the memory properly now

codenulls committed:
[FIX] Memory of dummy sequences' frames was not being freed for IFP Version 2. Now, it's being freed

codenulls committed:
added GetGateWayAnimationName and GetGateWayBlockName. Renamed PedClump to PedPointer of m_mapofPedPointers' functions

codenulls committed:
added new functions isNextAnimationCustom, setNextAnimationCustom, and setNextAnimationNormal

codenulls committed:
Modified setPedAnimation to play custom animation

codenulls committed:
IFP pointer is inserted to map if IFP is loaded sucessfully

codenulls committed:
added new functions for adding IFP pointer to map and checks for playing custom animation, not complete though. Custom animation is checked nicely, but isGateWayAnimationHierarchy function is not working properly.

codenulls committed:
custom IFP pointer is removed from map container when unloaded.

codenulls committed:
remove PedClump map functions to PedPointer map functions. Also added GetWayBlockName and GetGateWayAnimationName

codenulls committed:
added a boolean for checking next animation whether it's custom or not in CClientPed::CClientPed. Renamed PedClump function calls to PedPointer functions

codenulls committed:
added IFP function prototypes and a map container for retrieving IFP pointer by custom IFP block name

codenulls committed:
Merge pull request #1 from saml1er/custom_ifp_animations

Custom ifp animations

codenulls committed:
changed BlendAnimation hook to BlendAnimation_Hierarchy and hooks works fine. It is called when you execute setPedAnimation from client lua. This is where we can play custom animation

codenulls committed:
fixed bRunningCompressed in CAnimBlendHierarchySAInterface

codenulls committed:
change pAnimBlendNodesSequenceArray type from int * to DWORD * in CAnimBlendStaticAssociationSA

codenulls committed:
pInterface variable is initialized to nullptr in GetAnimStaticAssociation

codenulls committed:
[FIX] Fixed custom animations crashing by changing keyframes memory allocator from malloc to OLD_CMemoryMgr_Malloc which is GTA SA function

codenulls committed:
added a Name member to IFP_Animation struct for keeping the animation name

codenulls committed:
BlendAnimation hook has been changed to BlendAnimation_Hierarchy hook because that is only triggered when setPedAnimation is called. Commented AddAnimation hook for now, it can be used for engineRestoreAnimation and engineReplaceAnimation along with AddAnimationAndSync hook

codenulls committed:
changed BlendAnimationHandler hook resturn type

codenulls committed:
setPedAnimation finally works for playing custom animations. Both GTA VC and SA have been tested.

codenulls committed:
[FIX] unloadIFP function works good now.

codenulls committed:
Block name is now checked when loading IFP.
Before you could load many IFPs with same block name and it would cause memory leaks, now it's fixed.

codenulls committed:
Removed CClientPed pointer map functions

codenulls committed:
[BUGGY] Added a new function GetclientPedByClump, but causes crash sometimes

codenulls committed:
Added new functions for finding CClientPed by clump and used it in BlendAnimationHandler

codenulls committed:
Fixes in ModelRequestCallBack and CClientPed constructor and destructor.
Removed code for inserting and remove clump from map in ModelRequestCallBack. Also, CClientPed is inserted into map in constructor call and removed when destructor is called.

codenulls committed:
Animation name is now checked in engineLoadIFP whether the hierarchy exists or not

codenulls committed:
enabled AddAnimation hook

codenulls committed:
Removed CClient Ped pointer map functions

codenulls committed:
Merge pull request #2 from saml1er/custom_ifp_animations

Working custom IFP animations

codenulls committed:
[FIX] Fixed AddAnimation hook and disabled AddAnimationAndSync hook.
I used Original code for AddAnimation hook in CClientGame::AddAnimationHandler. It works well. Now, it's time to create our own CAnimBlendAssociationSAInterface instead of using CopyAnimation. I'll need to enable AddAnimationAndSync hook as well because this is called for almost every internal GTA animation.

codenulls committed:
Rewrote animation hooks properly in CClientGame and CMultiplayerSA
I renamed BlendAnimation hook to BlendAnimationHierarchy because now it's hooking a different BlendAnimation function that takes an animation hierarchy as argument. Also, I separated AddAnimation and AddAnimationAndSync hooks and their handlers. Now, they work independently. Also, there was a problem with AddAnimationAndSync hook in CMultiplayerSA.cpp, I fixed the return of the hook.

codenulls committed:
Removed CopyAnimation hook in InitHooks_FixBadAnimId

codenulls committed:
Removed code from AddAnimationHandler and AddAnimationAndSyncHandler
The code worked perfectly except for custom CopyAnimation function, if we tried to rewrite this function, it would crash.

codenulls committed:
Added ability to replace GTA internal animations.
Modified AddAnimation, AddAnimationAndSync, and CopyAnimation for replacing GTA internal animations with custom animations. We'll need to create our own static association when engineReplaceAnimation is called and use that static association for a specific ped/player to play our custom animation instead.

codenulls committed:
optimized CopyAnimation hook for saving tons of RAM
If we created a new static association for every player for every single animation then it means for 1000 players, the memory might exceed 400mb, that's a lot. Why not destroy the static association when we copy the animation from it? That's exactly what's happening in the hook now.

codenulls committed:
Added constexpr getAnimAssocGroupInterface function.

codenulls committed:
Created a new AssocGroupCopyAnimation handler

codenulls committed:
added typedef for CAnimBlendStaticAssociation::FreeSequenceArray

codenulls committed:
Static association is created and destroyed within CopyAnimation hook

codenulls committed:
Static association is initialized within AssocGroupCopyAnimationHandler
The typedef "hCAnimBlendStaticAssociation_Init" needs to deleted for production code. We need to properly use factory method.

codenulls committed:
No longer using global variables in animation hooks

codenulls committed:
Merge pull request #3 from saml1er/custom_ifp_animations

Custom ifp animations

codenulls committed:
Added missing members to CAnimBlockSAInterface

codenulls committed:
added missing member pLinkPtr to Hierarchy interface

codenulls committed:
added GetResourceIFPRoot inline function

codenulls committed:
IFP pointer is removed first from map then onClientIFPUnload is called

codenulls committed:
engineReplaceAnimation and engineRestoreAnimation is complete

codenulls committed:
Merge pull request #4 from saml1er/custom_ifp_animations

Custom ifp animations

codenulls committed:
added a check in GetAnimationHierarchy whether IFP is loaded or not

codenulls committed:
just added some padding, nothing special

codenulls committed:
added a new parameter and changed return type of AssocGroupCopyAnimation

Added SIFPAnimations ** as a new parameter and changed the return type to bool for both ( static and none static ) handlers of AssocGroupCopyAnimation

codenulls committed:
Animations are stored in SIFPAnimations struct now

codenulls committed:
added SIFPAnimations and removed a member from IFP_Animation

codenulls committed:
Added include guard and removed prototypes with CAnimBlock in them

codenulls committed:
changed CopyAnimation handler prototype and added SIFPAnimatins

codenulls committed:
[WIP] modifed CopyAnimation hook for SIFPAnimations
This is needed for unloading IFP. We'll use this to increment and decrement iReferences in SIFPAnimations when an animation association is created and destroyed respectively.

codenulls committed:
Moved custom animation hooks to a separate file

codenulls committed:
Added CAnimBlendAssociation Contructor and Destructor handler
This is for properly unloading IFP using SIFPAnimations struct.

codenulls committed:
added GetIFPAnimationsPointer
This function will get a pointer to SIFPAnimations struct of the IFP

codenulls committed:
Changed getReplacedAnimation return type to SReplacedAnimation *

codenulls committed:
Fixed CopyAnim hook and added mapOfCustomAnimationAssociations and its functions.

codenulls committed:
CopyAnimation handler now modifies pOutIFPAnimations.

codenulls committed:
inlined functions and added IFPAnimationsPointer functions

codenulls committed:
custom animation associations are kept in map for setPedAnimation

codenulls committed:
added isIFPLoaded and s_iAnimationSearchReferences to CClientIFP

codenulls committed:
s_iAnimationSearchReferences is checked before unloading IFP anims

codenulls committed:
GetAnimationHierarchy loop breaks after hierarchy is found

codenulls committed:
changed UnloadIFPAnimations to DeleteIFPAnimations

codenulls committed:
ReplaceAnimation checks whether IFP is loaded or not

codenulls committed:
Removed static for iAnimationSearchReferences

codenulls committed:
UnloadIFP calls DeleteIFPAnimations now and renamed iAnimSearchRef

codenulls committed:
Fixed unload crash. IFP is unloaded by checking references now.

codenulls committed:
Removed iAnimationSearchReferences.

codenulls committed:
Completed BlendAnimationHierarchy and CAnimBlendAssoc_Constructor
These two hooks are meant to work in the same way. We can use anyone we wish to. BlendAnimationHierarchy is much cleaner and easy to understand, so let's use that one. Keep in mind that these two hooks get triggered for setPedAnimation only. They have nothing to do with AddAnimation and AddAniamtionAndSync

codenulls committed:
Removed CAnimBlendAssoc_Hierarchy_Constructor hook and handler

codenulls committed:
Merge pull request #5 from saml1er/custom_ifp_animations

Custom ifp animations [Fixed unloading]

codenulls committed:
Added functions for CAnimBlendSequence and SA

codenulls committed:
Added GetAnimBlendSequence

codenulls committed:
added GetAnimBlendSequence

codenulls committed:
Used Factory pattern for CAnimBlendSequenceSAInterface

codenulls committed:
Added animation hierarchy functions

codenulls committed:
Removed destructor

codenulls committed:
Added GetCustomAnimBlendHierarchy and GetCustomAnimBlendSequence

codenulls committed:
Added GetCustomAnimBlendHierarchy and GetCustomAnimBlendSequence

codenulls committed:
Used factory pattern for Animation hierarchy and sequence

codenulls committed:
Added InsertAnimationDummySequence with new parameters

codenulls committed:
Added IS_IFP macro

codenulls committed:
Replaced raw pointer with shared_ptr for CClientIFP

codenulls committed:
Created CIFPAnimations class for keeping IFP animations only

codenulls committed:
Using shared_ptr for CIFPAnimations

codenulls committed:
Removed useless function calls in DeleteElementSpecial

codenulls committed:
Removed printf

codenulls committed:
Removed isIFPLoaded checks

codenulls committed:
Added IFP_Animation struct and removed unused members.

codenulls committed:
Moved structs and enums to CClientIFP class

codenulls committed:
Improved code readability. Still more improvement needed.

codenulls committed:
Improved key frames reading logic for IFP Version 2

codenulls committed:
used iterator in ReadIFPVersion2 instead of C loop

codenulls committed:
Added PraseSequenceVersion1 and PraseSequenceVersion2

codenulls committed:
Improved IFP Readers, both Version 1 and 2.

codenulls committed:
Improved CClientIFP design even more and some optimizations
The code design was really bad. Methods were very long which made them hard to understand. Now, they are readable.

codenulls committed:
Improved overall code quality. Methods are even more shorter now.

codenulls committed:
Removed IFPLoader include

codenulls committed:
Removed unused files

codenulls committed:
Moved CIFPAnimations to deathmatch/logic

codenulls committed:
Improved code design of CFileReader

codenulls committed:
Added new functions to CAnimManager for CClientIFP

codenulls committed:
Merge pull request #6 from saml1er/custom_ifp_animations

Custom ifp animations ( Improved Code Design )

codenulls committed:
Added CIFPEngine which contains most logic for Lua IFP functions

codenulls committed:
Code design improvement in CIFPAnimations::DeleteAnimations

codenulls committed:
Moved raw memory access to Game SA

codenulls committed:
[ FIX ] engineReplaceAnimation fails if block is not loaded

codenulls committed:
Optimized std::map key for performance.

codenulls committed:
[FIX] Unloading IFP will allow gateway animation to play instead of custom
Custom animation was not playing if you unloaded IFP. I used a shared_ptr to fix this problem. Also, unloading IFP resulted in looped animations to play gateway animation instead of custom, so fixed that as well.

codenulls committed:
Moved Get functions to header in CAnimManagerSA

codenulls committed:
Implemented thread synchronization for maps and set

codenulls committed:
Improved code design for CAnimBlendStaticAssociation

codenulls committed:
Replaced Animations map is now thread-safe in CClientPed

codenulls committed:
Removed unsed GetClientPedByClump from CClientEntity

codenulls committed:
Removed 2 unused functions in CGame and CGameSA

codenulls committed:
Merge pull request #7 from saml1er/custom_ifp_animations

Custom ifp animations ( Everything is complete )

codenulls committed:
[ FIX ] Compiling in release mode caused crash on CAnimBlendAssoc_destructor

codenulls committed:
Used Hungarian notation for naming structs and enums

codenulls committed:
Disabled console and removed printf

codenulls committed:
Merge pull request #8 from saml1er/custom_ifp_animations

Custom ifp animations ( mergeable to master )

codenulls committed:
Fix engineIFPLoad failure on resource multiple restarts

IFP was failing to load after restarting resource multiple times, because it was removed from map in CElementDeleter::DeleteIFP which caused a small delay in ms. Now, it will be removed from map when it's added to element deleter list, so engineLoadIFP won't fail because the element with the same block name does not exist in map anymore.

codenulls committed:
Merge branch 'test_custom_ifp_animations'

# Conflicts:
#	Client/game_sa/CAnimBlendAssocGroupSA.cpp
#	Client/game_sa/CAnimBlendAssocGroupSA.h
#	Client/game_sa/CAnimBlendHierarchySA.cpp
#	Client/game_sa/CAnimBlendHierarchySA.h
#	Client/game_sa/CAnimBlendSequenceSA.h
#	Client/game_sa/CAnimBlendStaticAssociationSA.cpp
#	Client/game_sa/CAnimBlendStaticAssociationSA.h
#	Client/game_sa/CAnimBlockSA.h
#	Client/game_sa/CAnimManagerSA.cpp
#	Client/game_sa/CAnimManagerSA.h
#	Client/mods/deathmatch/logic/CClientGame.cpp
#	Client/mods/deathmatch/logic/CClientGame.h
#	Client/mods/deathmatch/logic/CClientPed.cpp
#	Client/mods/deathmatch/logic/CClientPed.h
#	Client/mods/deathmatch/logic/CElementDeleter.cpp
#	Client/mods/deathmatch/logic/CElementDeleter.h
#	Client/mods/deathmatch/logic/CResource.cpp
#	Client/mods/deathmatch/logic/CResource.h
#	Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp
#	Client/mods/deathmatch/logic/CStaticFunctionDefinitions.h
#	Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp
#	Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.h
#	Client/multiplayer_sa/CMultiplayerSA.cpp
#	Client/multiplayer_sa/CMultiplayerSA.h
#	Client/multiplayer_sa/CMultiplayerSA_FixBadAnimId.cpp
#	Client/sdk/game/CAnimBlendAssocGroup.h
#	Client/sdk/game/CAnimBlendHierarchy.h
#	Client/sdk/game/CAnimBlock.h
#	Client/sdk/game/CAnimManager.h
#	Client/sdk/game/CGame.h
#	Client/sdk/multiplayer/CMultiplayer.h

codenulls committed:
Fix source formatting

Commit:
Merge branch 'master' into custom_ifp_animations

codenulls committed:
Restored CStaticFunctionDefinitions::GUIMoveToBack

Three lines got removed from this function while solving conflicts, so restored them back.

sbx320 commented on Client/mods/deathmatch/logic/CClientGame.cpp:
```diff
@@ -6659,3 +6770,111 @@ void CClientGame::RestreamModel(unsigned short usModel)
+{
+    std::lock_guard<std::mutex> mutexGuardedLock(m_MutexOfPedPointersSet);
+    IFP->MarkAsUnloading();
+    for (auto it = m_setOfPedPointers.begin(); it != m_setOfPedPointers.end(); it++)
```
Why not `for (auto& it : m_setOfPedPointers)` ?

sbx320 commented on Client/mods/deathmatch/logic/CClientIFP.h:
```diff
@@ -0,0 +1,288 @@
+#pragma once
+
+#ifndef __CCLIENTIFP_H
```
`#pragma once` and include guards combined are a bit redundant

sbx320 commented on Client/mods/deathmatch/logic/CElementDeleter.cpp:
```diff
@@ -72,6 +76,32 @@ void CElementDeleter::DeleteRecursive(class CClientEntity* pElement)
+{
+    CClientIFP&        IFP = static_cast<CClientIFP&>(*pElement);
+    const unsigned int u32BlockNameHash = IFP.GetBlockNameHash();
+    for (auto it = m_vecIFPElements.begin(); it != m_vecIFPElements.end(); ++it)
```
I'd go with `std::find_if` & `std::erase` here

sbx320 commented on Client/mods/deathmatch/logic/CFileReader.cpp:
```diff
@@ -0,0 +1,142 @@
+    if (!LoadToMemory())
+        return FALSE;
+
+    // std::cout << "Length: " << u32FileLength << std::endl;
```
;)

codenulls committed:
Rewrote CFileReader and some code review improvements

CFileReader is using ifstream to load file into std::vector. No more dynamic memory allocation.
&
replaced std::vector::push_back with std::vector::emplace_back in CElementDeleter::OnClientIFPElementDestroy
& Removed empty destructor from CClientIFP and empty constructor from CIFPAnimations
- Suggested by Necktrox

Used for (auto& pPed : m_setOfPedPointers) for iterating in CClientGame::OnClientIFPUnload.
&
using std::find_if in CElementDeleter::DeleteIFP with a lambda function
- Suggested by sbx320

Removed #pragma once in CClientIFP.h and CIFPAnimations.h
- Suggested by sbx320 and Necktrox

codenulls committed:
Code quality improvements

Replaced uint32_t, int32_t, and int16_t with std::* from <cstdint>
- Suggested by Necktrox

Used std::find_if in CClientIFP::GetAnimationHierarchy
& improved iterator loops by using
for (auto& Animation : m_pIFPAnimations->vecAnimations)  in CClientIFP::ReadIFPVersion2 and CClientIFP::ReadIFPVersion1
- earlier suggested by sbx320

Replaced std::string with SString.

Changed CIFPAnimations::SIFPAnimation::pSequencesMemory type to BYTE*. Removed casting where this member of struct was accessed.

Used const where applicable in CClientIFP.cpp.

Renamed ConvertStringToMapKey to ConvertStringToKey.

Applied source formatting again

ghost commented:
Something important about engineReplaceAnimation:

I forgot to mention that partial animations are not supported for replacing, if you replace them, MTA won't crash, but the animation will get glitched, most probably not what you want. Those are animations which usually run parallel with other animations like punching while running can't be replaced. You can replace running, punching, crouching, shooting with weapons, swimming, and hundreds of more similar animations, but there are few that you can't replace properly, and those few animations are partial animations.

The only reason why I did not fix this is because it will require me to rewrite the entire animation system including GTA task system. The amount of bones to be moved for every animation is hardcoded in GTA:SA for every task, and it will literally take me 4-5 more months to completely reverse it, and a few more to write code properly for MTA. That's roughly 1 year :) 

I do have plans to add support for partial animations and the ability to have any amount of bones. Currently, bone limit is 32.  I want to lift this limit and add a few more functions like setPedBonePosition and setPedBoneRotation. Hopefully, next year.

jushar commented on Client/mods/deathmatch/logic/CClientGame.cpp:
```diff
@@ -11,6 +11,9 @@
 #include "StdInc.h"
 #include <net/SyncStructures.h>
+#include <../Client/game_sa/CAnimBlendAssocGroupSA.h>
+#include <../Client/game_sa/CAnimBlendAssociationSA.h>
```
GameSA headers should not be used here. Instead, add an interface to the Game SDK (/Client/sdk/game).

codenulls committed:
Removed GameSA includes from Client Deathmatch and minor code improvements

Replaced GameSA includes with interfaces
- Suggested by Jusonex

Removed #pragma from CIFPEngine.h
eRestoreType is now passed by const reference in CIFPEngine::EngineRestoreAnimation

Pirulax commented:
Look pretty nice, this is what i was waiting for, thanks for it, u did an awesome job, even do, i never wrote a single line of C++ :D

codenulls committed:
Removed IFP mutexes from CClientGame and CClientPed

They were not needed. I used them because I was in doubt whether GTA:SA is using a secondary thread. I confirmed that it's not using another thread with the help of SharedUtil::IsMainThread() method.
- Suggested by Jusonex

codenulls committed:
Improved custom animation hooks for readability

Rewrote HOOK_CAnimBlendAssocGroup_CopyAnimation in C++ for more clarity, it is shorter now, and thus more readable.

Moved HOOK_CAnimBlendAssoc_destructor to CMultiplayerSA_HookDestructors.cpp

codenulls committed:
Fix warning c4150 when deleting Game SA interface

Visual studio was generating a warning when calling delete operator on hierarchy and sequnece interface in Client Deathmatch, so replaced it with CAnimManager::DeleteCustomAnimHierarchyInterface and CAnimManager::DeleteCustomAnimSequenceInterface in CIFPAnimations::DeleteAnimations and CClientIFP::MoveSequencesWithDummies

Also, applied source formatting

neves768 commented:
Thx for your hard work on it, @codenulls and @saml1er. You're building a new high-attractive thing to MTA:SA. I hope it gets into the game soon. Also, thanks for the two reviewers that are taking their times to analyze and test it.  :heart:

botder commented on Client/mods/deathmatch/logic/CClientIFP.cpp:
```diff
@@ -0,0 +1,1079 @@
+#include <StdInc.h>
```
File is missing the copyright/etc. comment at the top

qaisjp replied:
Use the header from here: https://github.com/multitheftauto/mtasa-blue/blob/master/Shared/mods/deathmatch/logic/luadefs/CLuaBitDefs.cpp

botder commented on Client/mods/deathmatch/logic/CClientIFP.cpp:
```diff
@@ -0,0 +1,1079 @@
+        if (bAnp2 || bAnp3)
+        {
+            m_bVersion1 = false;
+            ReadIFPVersion2(bAnp3);
```
What if the IFP file is corrupt? Shouldn't this function call return a boolean to indicate failure? See also ReadIFPVersion1

botder commented on Client/mods/deathmatch/logic/CClientIFP.h:
```diff
@@ -0,0 +1,285 @@
+#ifndef __CCLIENTIFP_H
```
File is missing the copyright/etc. comment at the top.
Also, use **#pragma once**

botder commented on Client/mods/deathmatch/logic/CFileReader.cpp:
```diff
@@ -0,0 +1,51 @@
+#include <StdInc.h>
```
File is missing the copyright/etc. comment at the top

botder commented on Client/mods/deathmatch/logic/CFileReader.h:
```diff
@@ -0,0 +1,36 @@
+#ifndef CFILEREADER_H
```
File is missing the copyright/etc. comment at the top.
Also, use **#pragma once**

botder commented on Client/mods/deathmatch/logic/CIFPAnimations.cpp:
```diff
@@ -0,0 +1,38 @@
+#include <StdInc.h>
```
File is missing the copyright/etc. comment at the top

botder commented on Client/mods/deathmatch/logic/CIFPAnimations.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef __CIFPANIMATIONS_H
```
File is missing the copyright/etc. comment at the top.
Also, use **#pragma once**

botder commented on Client/mods/deathmatch/logic/CIFPEngine.cpp:
```diff
@@ -0,0 +1,91 @@
+#include <StdInc.h>
```
File is missing the copyright/etc. comment at the top

botder commented on Client/mods/deathmatch/logic/CIFPEngine.h:
```diff
@@ -0,0 +1,21 @@
+#ifndef __CIFPENGINE_H
```
File is missing the copyright/etc. comment at the top.
Also, use **#pragma once**

botder commented on Client/multiplayer_sa/CMultiplayerSA_CustomAnimations.cpp:
```diff
@@ -0,0 +1,283 @@
+#include "StdInc.h"
```
File is missing the copyright/etc. comment at the top

botder commented on Client/multiplayer_sa/CMultiplayerSA_FixBadAnimId.cpp:
```diff
@@ -111,6 +119,6 @@ void _declspec(naked) HOOK_GetAnimHierarchyFromSkinClump()
 void CMultiplayerSA::InitHooks_FixBadAnimId(void)
 {
-    EZHookInstall(CAnimBlendAssocGroupCopyAnimation);
+    // EZHookInstall ( CAnimBlendAssocGroupCopyAnimation );
```
Is there any reason not to remove this line completely?

codenulls committed:
Added m_ prefix before member names and copyright comment at top.

codenulls committed:
Added copyright comment, and added #pragma once to header files

qaisjp committed:
Undo whitespace change

(test commit)

Commit:
Merge pull request #1 from multitheftauto/master

Merge master into custom_ifp_animations

jushar commented:
See https://github.com/multitheftauto/mtasa-blue/commit/15f99b2fcf70b8d5b82883c49cb2413cd39dee4b
